File: databases/sqlite/python/src/recipe_transactional_storage.py
"""
Recipe: Transactional Local Storage
Database: SQLite
Description: Demonstrates CRUD operations (create table, insert, update, select, delete)
             with WAL mode enabled, explicit transactions, and error handling.

Usage: python src/recipe_transactional_storage.py
"""

import sqlite3
import os
import sys


def main() -> None:
    db_path = os.path.join(os.path.dirname(__file__), "..", "recipe_storage.db")
    db_path = os.path.abspath(db_path)

    # ── 1. Setup ──────────────────────────────────────────────────────
    conn = sqlite3.connect(db_path)
    conn.execute("PRAGMA journal_mode=WAL")
    conn.execute("PRAGMA foreign_keys=ON")

    print(f"Connected to {db_path} (WAL mode)")

    # ── 2. Schema ─────────────────────────────────────────────────────
    conn.execute("""
        CREATE TABLE IF NOT EXISTS books (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            title TEXT NOT NULL,
            author TEXT NOT NULL,
            year INTEGER,
            created_at TEXT DEFAULT (datetime('now'))
        )
    """)
    conn.commit()
    print("Table 'books' ready.")

    # ── 3. Write: Insert ──────────────────────────────────────────────
    try:
        with conn:
            conn.execute(
                "INSERT INTO books (title, author, year) VALUES (?, ?, ?)",
                ("The Pragmatic Programmer", "Andy Hunt & Dave Thomas", 1999),
            )
            conn.execute(
                "INSERT INTO books (title, author, year) VALUES (?, ?, ?)",
                ("Designing Data-Intensive Applications", "Martin Kleppmann", 2017),
            )
        print("Inserted 2 books.")
    except sqlite3.Error as e:
        print(f"Insert error: {e}", file=sys.stderr)

    # ── 4. Read: Select ───────────────────────────────────────────────
    rows = conn.execute("SELECT id, title, author, year FROM books").fetchall()
    print(f"\nAll books ({len(rows)}):")
    for row in rows:
        print(f"  [{row[0]}] {row[1]} by {row[2]} ({row[3]})")

    # ── 5. Update ─────────────────────────────────────────────────────
    try:
        with conn:
            conn.execute(
                "UPDATE books SET year = ? WHERE title = ?",
                (2020, "The Pragmatic Programmer"),
            )
        print("\nUpdated year for 'The Pragmatic Programmer'.")
    except sqlite3.Error as e:
        print(f"Update error: {e}", file=sys.stderr)

    # ── 6. Verify update ──────────────────────────────────────────────
    row = conn.execute(
        "SELECT title, year FROM books WHERE title = ?",
        ("The Pragmatic Programmer",),
    ).fetchone()
    print(f"  Verified: {row[0]} -> year {row[1]}")

    # ── 7. Delete ─────────────────────────────────────────────────────
    try:
        with conn:
            conn.execute(
                "DELETE FROM books WHERE title = ?",
                ("Designing Data-Intensive Applications",),
            )
        print("\nDeleted 'Designing Data-Intensive Applications'.")
        remaining = conn.execute("SELECT COUNT(*) FROM books").fetchone()[0]
        print(f"  Remaining books: {remaining}")
    except sqlite3.Error as e:
        print(f"Delete error: {e}", file=sys.stderr)

    # ── 8. Cleanup ────────────────────────────────────────────────────
    conn.close()
    os.remove(db_path)
    print(f"\nCleaned up {db_path}.")


if __name__ == "__main__":
    main()

File: databases/sqlite/rust/src/bin/recipe_transactional_storage.rs
//! Recipe: Transactional Local Storage
//! Database: SQLite
//! Description: Demonstrates CRUD operations (create table, insert, update, select, delete)
//!              with WAL mode enabled, explicit transactions, and error handling.
//!
//! Usage: cargo run --bin recipe_transactional_storage

use rusqlite::{params, Connection, Result};
use std::fs;

fn main() -> Result<()> {
    let db_path = "recipe_storage.db";

    // ── 1. Setup ──────────────────────────────────────────────────────
    let conn = Connection::open(db_path)?;
    conn.execute_batch("PRAGMA journal_mode=WAL")?;
    conn.execute_batch("PRAGMA foreign_keys=ON")?;

    println!("Connected to {} (WAL mode)", db_path);

    // ── 2. Schema ─────────────────────────────────────────────────────
    conn.execute(
        "CREATE TABLE IF NOT EXISTS books (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            title TEXT NOT NULL,
            author TEXT NOT NULL,
            year INTEGER,
            created_at TEXT DEFAULT (datetime('now'))
        )",
        [],
    )?;
    println!("Table 'books' ready.");

    // ── 3. Write: Insert ──────────────────────────────────────────────
    conn.execute(
        "INSERT INTO books (title, author, year) VALUES (?1, ?2, ?3)",
        params!["The Pragmatic Programmer", "Andy Hunt & Dave Thomas", 1999],
    )?;
    conn.execute(
        "INSERT INTO books (title, author, year) VALUES (?1, ?2, ?3)",
        params!["Designing Data-Intensive Applications", "Martin Kleppmann", 2017],
    )?;
    println!("Inserted 2 books.");

    // ── 4. Read: Select ───────────────────────────────────────────────
    {
        let mut stmt = conn.prepare("SELECT id, title, author, year FROM books")?;
        let rows: Vec<(i64, String, String, i64)> = stmt
            .query_map([], |row| {
                Ok((row.get(0)?, row.get(1)?, row.get(2)?, row.get(3)?))
            })?
            .filter_map(|r| r.ok())
            .collect();

        println!("\nAll books ({}):", rows.len());
        for (id, title, author, year) in &rows {
            println!("  [{}] {} by {} ({})", id, title, author, year);
        }
    } // stmt dropped here, releasing borrow on conn

    // ── 5. Update ─────────────────────────────────────────────────────
    conn.execute(
        "UPDATE books SET year = ?1 WHERE title = ?2",
        params![2020, "The Pragmatic Programmer"],
    )?;
    println!("\nUpdated year for 'The Pragmatic Programmer'.");

    // ── 6. Verify update ──────────────────────────────────────────────
    let (title, year): (String, i64) = conn.query_row(
        "SELECT title, year FROM books WHERE title = ?1",
        params!["The Pragmatic Programmer"],
        |row| Ok((row.get(0)?, row.get(1)?)),
    )?;
    println!("  Verified: {} -> year {}", title, year);

    // ── 7. Delete ─────────────────────────────────────────────────────
    conn.execute(
        "DELETE FROM books WHERE title = ?1",
        params!["Designing Data-Intensive Applications"],
    )?;
    println!("\nDeleted 'Designing Data-Intensive Applications'.");
    let remaining: i64 =
        conn.query_row("SELECT COUNT(*) FROM books", [], |row| row.get(0))?;
    println!("  Remaining books: {}", remaining);

    // ── 8. Cleanup ────────────────────────────────────────────────────
    drop(conn); // Close the connection first
    fs::remove_file(db_path).expect("Failed to remove test database file");
    println!("\nCleaned up {}.", db_path);

    Ok(())
}
